sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -961,8 +961,8 @@ object TypeCoercion extends TypeCoercionBase {
 
       // Implicit cast between date time types
       case (DateType, TimestampType) => TimestampType
-      case (DateType, AnyTimestampType) => TimestampType
-      case (TimestampType, DateType) => DateType
+      case (DateType, AnyTimestampType) => AnyTimestampType.defaultConcreteType
```

Comment:
This addresses comment in https://github.com/apache/spark/pull/33136#discussion_r661468789